### PR TITLE
[6918] Correct msiAddKeyVal behavior (4-2-stable)

### PR DESCRIPTION
**Pull request by MartinFlores751**
`msiAddKeyValToMspStr()` now can handle empty strings/key-value pairs properly.

### CODE REVIEW

File: lib/core/src/msParam.cpp [hunk 1]
```diff
@@ -1451,60 +1451,61 @@ int
...
+    const char *valPtr{parseMspForStr(valStr)};
+    int valLen{static_cast<int>(get_length(valPtr))};
+
+    if (valLen + keyLen <= 0) {
         return 0;
     }
 
-    char *oldKeyValPtr = parseMspForStr( msKeyValStr );
+    char *oldKeyValPtr{parseMspForStr(msKeyValStr)};
     char *newKeyValPtr, *tmpPtr;
```
Comment (trel) [inline lib/core/src/msParam.cpp:1480]: split these up?   initialize?

File: lib/core/src/msParam.cpp [hunk 2]
```diff
@@ -1451,60 +1451,61 @@ int
...
-    if ( oldKeyValPtr == NULL ) {
-        int newLen = valLen + keyLen + 10;
-        newKeyValPtr = ( char * )malloc( newLen );
+    if (oldKeyValPtr == nullptr) {
+        const int newLen{valLen + keyLen + 10};
```
Comment (trel) [inline lib/core/src/msParam.cpp:1482]: what is 10?

File: lib/core/src/msParam.cpp [hunk 3]
```diff
@@ -1451,60 +1451,61 @@ int
...
-    }
-
-    if ( keyLen > 0 ) {
-        snprintf( tmpPtr, keyLen + 2, "%s=", keyPtr );
+        const int oldLen{static_cast<int>(std::strlen(oldKeyValPtr))};
+        const int newLen{oldLen + valLen + keyLen + 10};
+        newKeyValPtr = reinterpret_cast<char*>(std::malloc(newLen));
+	if (oldLen > 0) {
+	    std::snprintf(newKeyValPtr, newLen, "%s%s", oldKeyValPtr, MS_INP_SEP_STR);
+	    tmpPtr = newKeyValPtr + oldLen + 4;
```
Comment (trel) [inline lib/core/src/msParam.cpp:1493]: what is 4?

File: lib/core/src/msParam.cpp [hunk 4]
```diff
@@ -1451,60 +1451,61 @@ int
...
+	} else {
+  	    *newKeyValPtr = '\0';
+	    tmpPtr = newKeyValPtr;
+	}
+	std::free(oldKeyValPtr);
+    }
+
+    if (keyLen > 0) {
+        std::snprintf(tmpPtr, keyLen + 2, "%s=", keyPtr);
```
Comment (trel) [inline lib/core/src/msParam.cpp:1502]: what is 2?

**Review by trel — COMMENTED** [review]
so these tests failed before the code edit?  because of the ++++?

File: scripts/irods/test/test_all_rules.py [hunk 5]
```diff
@@ -663,6 +663,82 @@ def test_msiServerMonPerf_default_3736(self):
 
         self.rods_session.assert_icommand("irule -F " + rule_file);
 
+    @unittest.skipUnless(plugin_name == 'irods_rule_engine_plugin-irods_rule_language', 'only applicable for irods_rule_language REP')
+    def test_msiAddKeyValToMspStr_works_with_empty_string_6918(self):
```
Comment (korydraughn) [inline scripts/irods/test/test_all_rules.py:667]: `_6918` -> `__issue_6918`

File: scripts/irods/test/test_all_rules.py [hunk 6]
```diff
@@ -663,6 +663,82 @@ def test_msiServerMonPerf_default_3736(self):
...
+'''.format(collection=collection_to_free)
+
+        with open(rule_file, 'w') as f:
+            f.write(rule_string)
+
+        rep_name = 'irods_rule_engine_plugin-irods_rule_language-instance'
+        self.admin.assert_icommand(['irule', '-r', rep_name, '-F', rule_file])
+
+    @unittest.skipUnless(plugin_name == 'irods_rule_engine_plugin-irods_rule_language', 'only applicable for irods_rule_language REP')
+    def test_msiRmColl_removes_collection_via_flag_6918(self):
```
Comment (korydraughn) [inline scripts/irods/test/test_all_rules.py:718]: `_6918` -> `__issue_6918`

File: lib/core/src/msParam.cpp [hunk 7]
```diff
@@ -1451,60 +1451,61 @@ int
...
-    }
-    else {
-        keyLen = strlen( keyPtr );
-    }
+    auto get_length{[](const char* str) -> size_t {
+        if (str == nullptr || std::strcmp(str, MS_NULL_STR) == 0) {
+	    return 0;
+	}
+	return std::strlen(str);
+    }};
```
Comment (korydraughn) [inline lib/core/src/msParam.cpp:1467]: Formatting.

File: lib/core/src/msParam.cpp [hunk 8]
```diff
@@ -1451,60 +1451,61 @@ int
...
 
-    char *keyPtr = parseMspForStr( keyStr );
-    int keyLen;
-    if ( keyPtr == NULL || strcmp( keyPtr, MS_NULL_STR ) == 0 ) {
-        keyLen = 0;
-    }
-    else {
-        keyLen = strlen( keyPtr );
-    }
+    auto get_length{[](const char* str) -> size_t {
```
Comment (korydraughn) [inline lib/core/src/msParam.cpp:1462]: `size_t` -> `std::size_t`

File: lib/core/src/msParam.cpp [hunk 9]
```diff
@@ -1451,60 +1451,61 @@ int
...
 
-    char *oldKeyValPtr = parseMspForStr( msKeyValStr );
+    char *oldKeyValPtr{parseMspForStr(msKeyValStr)};
     char *newKeyValPtr, *tmpPtr;
-    if ( oldKeyValPtr == NULL ) {
-        int newLen = valLen + keyLen + 10;
-        newKeyValPtr = ( char * )malloc( newLen );
+    if (oldKeyValPtr == nullptr) {
+        const int newLen{valLen + keyLen + 10};
+        newKeyValPtr = reinterpret_cast<char*>(std::malloc(newLen));
```
Comment (korydraughn) [inline lib/core/src/msParam.cpp:1483]: `reinterpret_cast` -> `static_cast`

For correctness: `newLen` -> `sizeof(char) * newLen`

File: lib/core/src/msParam.cpp [hunk 10]
```diff
@@ -1451,60 +1451,61 @@ int
...
-        newKeyValPtr = ( char * )malloc( newLen );
-        snprintf( newKeyValPtr, newLen, "%s%s", oldKeyValPtr, MS_INP_SEP_STR );
-        tmpPtr = newKeyValPtr + oldLen + 4;
-        free( oldKeyValPtr );
-    }
-
-    if ( keyLen > 0 ) {
-        snprintf( tmpPtr, keyLen + 2, "%s=", keyPtr );
+        const int oldLen{static_cast<int>(std::strlen(oldKeyValPtr))};
+        const int newLen{oldLen + valLen + keyLen + 10};
```
Comment (korydraughn) [inline lib/core/src/msParam.cpp:1489]: What is 10?

File: lib/core/src/msParam.cpp [hunk 11]
```diff
@@ -1451,60 +1451,61 @@ int
...
+        newKeyValPtr = reinterpret_cast<char*>(std::malloc(newLen));
```
Comment (korydraughn) [inline lib/core/src/msParam.cpp:1490]: - static_cast
- sizeof * newLen

File: lib/core/src/msParam.cpp [hunk 12]
```diff
@@ -1451,60 +1451,61 @@ int
...
+        std::snprintf(tmpPtr, keyLen + 2, "%s=", keyPtr);
         tmpPtr += keyLen + 1;
     }
-    if ( valLen > 0 ) {
-        snprintf( tmpPtr, valLen + 2, "%s", valPtr );
+    if (valLen > 0) {
+        std::snprintf(tmpPtr, valLen + 2, "%s", valPtr);
     }
-    msKeyValStr->inOutStruct = ( void * ) newKeyValPtr;
+    msKeyValStr->inOutStruct = reinterpret_cast<void *>(newKeyValPtr);
```
Comment (korydraughn) [inline lib/core/src/msParam.cpp:1508]: You can remove the cast here.

C++ allows pointers to be assigned to void pointers implicitly.

File: scripts/irods/test/test_all_rules.py [hunk 13]
```diff
@@ -663,6 +663,82 @@ def test_msiServerMonPerf_default_3736(self):
...
+        self.admin.assert_icommand(['itouch', collection_to_free])
+        self.admin.assert_icommand(['irm', collection_to_free])
+
+        rule_file="test_msiAddKeyValToMspStr.r"
+        rule_string= '''
+msiTestAddKeyValToMspStr {{
+    *Flag = "";
+    *Coll = "{trash_col}";
+    msiAddKeyValToMspStr("irodsAdminRmTrash", "", *Flag);
+    msiRmColl(*Coll, *Flag, *result);
```
Comment (korydraughn) [inline scripts/irods/test/test_all_rules.py:680]: How does L680 help prove `msiAddKeyValToMspStr` handles empty strings correctly?

File: scripts/irods/test/test_all_rules.py [hunk 14]
```diff
@@ -663,6 +663,82 @@ def test_msiServerMonPerf_default_3736(self):
...
+'''.format(trash_col=os.path.join(self.admin.session_collection_trash))
+
+        with open(rule_file, 'w') as f:
+            f.write(rule_string)
+
+        rep_name = 'irods_rule_engine_plugin-irods_rule_language-instance'
+        self.admin.assert_icommand(['irule', '-r', rep_name, '-F', rule_file])
+
+    @unittest.skipUnless(plugin_name == 'irods_rule_engine_plugin-irods_rule_language', 'only applicable for irods_rule_language REP')
+    def test_msiRmColl_removes_collection_6918(self):
```
Comment (korydraughn) [inline scripts/irods/test/test_all_rules.py:694]: `_6918` -> `__issue_6918`

File: lib/core/src/msParam.cpp [hunk 15]
```diff
@@ -1451,60 +1451,61 @@ int
...
-        valLen = strlen( valPtr );
-    }
-    if ( valLen + keyLen <= 0 ) {
+    const char *keyPtr{parseMspForStr(keyStr)};
+    int keyLen{static_cast<int>(get_length(keyPtr))};
+
+    const char *valPtr{parseMspForStr(valStr)};
+    int valLen{static_cast<int>(get_length(valPtr))};
+
+    if (valLen + keyLen <= 0) {
```
Comment (korydraughn) [inline lib/core/src/msParam.cpp:1475]: Consider adding some words explaining the purpose of this.

File: lib/core/src/msParam.cpp [hunk 16]
```diff
@@ -1451,60 +1451,61 @@ int
 addKeyValToMspStr( msParam_t * keyStr, msParam_t * valStr,
                    msParam_t * msKeyValStr ) {
 
-    if ( ( keyStr == NULL && valStr == NULL ) || msKeyValStr == NULL ) {
+    if ((keyStr == nullptr && valStr == nullptr) || msKeyValStr == nullptr) {
```
Comment (korydraughn) [inline lib/core/src/msParam.cpp:1454]: What happens if `valStr` is valid, but `keyStr` is null?
Do we protect against that case later in the function?

File: scripts/irods/test/test_all_rules.py [hunk 17]
```diff
@@ -663,6 +663,82 @@ def test_msiServerMonPerf_default_3736(self):
...
+        collection_to_free = os.path.join(self.admin.session_collection, 'freeMe')
+
+        self.admin.assert_icommand(['imkdir', collection_to_free])
+
+        rule_file="test_msiRmCollSimple.r"
+        rule_string= '''
+msiTestAddKeyValToMspStr {{
```
Comment (alanking) [inline scripts/irods/test/test_all_rules.py:701]: Let's update the name of the rule to reflect the test

**Review by alanking — COMMENTED** [review]
I'll throw my twig onto the pile :)

File: lib/core/src/msParam.cpp (hunk 1 above)
Comment (MartinFlores751) [reply lib/core/src/msParam.cpp:1480]: The `char *newKeyValPtr, *tmpPtr;` right? Can do!

File: lib/core/src/msParam.cpp (hunk 2 above)
Comment (MartinFlores751) [reply lib/core/src/msParam.cpp:1482]: Not sure myself, I'll try looking to see if there's something that might give a hint.

File: lib/core/src/msParam.cpp (hunk 7 above)
Comment (MartinFlores751) [reply lib/core/src/msParam.cpp:1467]: I don't see it, and I don't think 4.2 has clang format, so I can't cheat and use that. Are you saying the `if` should be split at the `||` or is it something else?

Comment (korydraughn) [reply lib/core/src/msParam.cpp:1467]: Perhaps your editor is hiding the issue.

Try looking at those lines in your browser.

Comment (korydraughn) [reply lib/core/src/msParam.cpp:1467]: The indentation is incorrect.

File: lib/core/src/msParam.cpp (hunk 3 above)
Comment (trel) [reply lib/core/src/msParam.cpp:1493]: or just use size_of (or whatever the correct equivalent is)...

File: lib/core/src/msParam.cpp (hunk 7 above)
Comment (trel) [reply lib/core/src/msParam.cpp:1467]: i see tabs, and we're not doing tabs in irods/irods i believe...

Comment (korydraughn) [reply lib/core/src/msParam.cpp:1467]: Correct.

File: scripts/irods/test/test_all_rules.py (hunk 17 above)
Comment (MartinFlores751) [reply scripts/irods/test/test_all_rules.py]: Should be done!

File: scripts/irods/test/test_all_rules.py (hunk 13 above)
Comment (MartinFlores751) [reply scripts/irods/test/test_all_rules.py:708]: Added test to check `msiAddKeyValToMspStr`, kept this one to test `msiRmColl`.

File: lib/core/src/msParam.cpp (hunk 12 above)
Comment (MartinFlores751) [reply lib/core/src/msParam.cpp]: Done! Didn't know about that implicit cast.

File: lib/core/src/msParam.cpp (hunk 7 above)
Comment (MartinFlores751) [reply lib/core/src/msParam.cpp]: Should be good, wouldn't be surprised if more formatting errors showed up though. Haven't properly set up my new editor yet.

File: lib/core/src/msParam.cpp (hunk 3 above)
Comment (MartinFlores751) [reply lib/core/src/msParam.cpp]: Name given! If too wordy, I'll change to something else.

File: lib/core/src/msParam.cpp (hunk 16 above)
Comment (MartinFlores751) [reply lib/core/src/msParam.cpp:1454]: I believe both `parseMspForStr()` and the lambda `get_length()` handle the `nullptr` case. Whether we want a value but no key is a different discussion: https://github.com/irods/irods/pull/7049#discussion_r1171576383.

File: scripts/irods/test/test_all_rules.py [hunk 18]
```diff
@@ -664,16 +664,44 @@ def test_msiServerMonPerf_default_3736(self):
...
+    def test_msiAddKeyValToMspStr_works_with_empty_string__issue_6918(self):
+        rule_file="test_msiAddKeyValToMspStr.r"
+        rule_string= '''
+msiTestAddKeyValToMspStr {{
+    *Flag = "";
+    msiAddKeyValToMspStr("keyOnly", "", *Flag);
+    writeLine("stdout", *Flag);
+
+    *Flag = "";
+    msiAddKeyValToMspStr("", "valueOnly", *Flag);
```
[the remark this answers is not in the document]
Comment (trel) [reply scripts/irods/test/test_all_rules.py:676]: pretty sure we don't want that to ever happen.  someone should get a big fat error if they try.

can't really think of a legitimate use case for this scenario...

consider returning an error - and see if the test suite falls over somewhere.

[the remark this answers is not in the document]
Comment (MartinFlores751) [reply scripts/irods/test/test_all_rules.py]: Should be removed, currently returns `USER_INPUT_STRING_ERR`. Let me know if you know something that works better for this case!

[the remark this answers is not in the document]
Comment (alanking) [reply scripts/irods/test/test_all_rules.py]: My only suggestion would be to try using the value-only MSI invocation in a test and ensure that the expected error comes back. I can't really find an error code that conveys things any better, so, seems fine to me.

[the remark this answers is not in the document]
Comment (alanking) [reply scripts/irods/test/test_all_rules.py]: After some discussion, we've decided that we should just leave the value-only behavior for now. Needs more investigation. Please resolve this once https://github.com/irods/irods/pull/7049/commits/f53942b90b95a1ea7df59f47fc01aad975894190 is reverted.

File: lib/core/src/msParam.cpp (hunk 2 above)
Comment (MartinFlores751) [reply lib/core/src/msParam.cpp]: Note: We're going to open a separate issue to investigate this magical number.

**Review by alanking — APPROVED** [review]
